exchanges/stream/stream_match.go

```diff
@@ -5,76 +5,69 @@ import (
 	"sync"
 )
 
+var (
+	errSignatureCollision = errors.New("signature collision")
+	errInvalidBufferSize  = errors.New("buffer size must be positive")
+)
+
 // NewMatch returns a new Match
 func NewMatch() *Match {
-	return &Match{
-		m: make(map[interface{}]chan []byte),
-	}
+	return &Match{m: make(map[any]incoming)}
 }
 
 // Match is a distributed subtype that handles the matching of requests and
 // responses in a timely manner, reducing the need to differentiate between
 // connections. Stream systems fan in all incoming payloads to one routine for
 // processing.
 type Match struct {
-	m  map[interface{}]chan []byte
+	m  map[any]incoming
 	mu sync.Mutex
 }
 
-// Matcher defines a payload matching return mechanism
-type Matcher struct {
-	C   chan []byte
-	sig interface{}
-	m   *Match
-}
-
-// Incoming matches with request, disregarding the returned payload
-func (m *Match) Incoming(signature interface{}) bool {
-	return m.IncomingWithData(signature, nil)
+type incoming struct {
+	expected int
+	c        chan<- []byte
 }
 
 // IncomingWithData matches with requests and takes in the returned payload, to
 // be processed outside of a stream processing routine and returns true if a handler was found
-func (m *Match) IncomingWithData(signature interface{}, data []byte) bool {
+func (m *Match) IncomingWithData(signature any, data []byte) bool {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 	ch, ok := m.m[signature]
-	if ok {
-		select {
-		case ch <- data:
-		default:
-			// this shouldn't occur but if it does continue to process as normal
-			return false
-		}
-		return true
+	if !ok {
+		return false
+	}
+	ch.c <- data
+	ch.expected--
+	if ch.expected == 0 {
+		close(ch.c)
+		delete(m.m, signature)
 	}
-	return false
+	return true
 }
 
 // Set the signature response channel for incoming data
-func (m *Match) Set(signature interface{}) (Matcher, error) {
-	var ch chan []byte
+func (m *Match) Set(signature any, bufSize int) (<-chan []byte, error) {
+	if bufSize <= 0 {
+		return nil, errInvalidBufferSize
+	}
 	m.mu.Lock()
+	defer m.mu.Unlock()
 	if _, ok := m.m[signature]; ok {
-		m.mu.Unlock()
-		return Matcher{}, errors.New("signature collision")
+		return nil, errSignatureCollision
 	}
-	// This is buffered so we don't need to wait for receiver.
-	ch = make(chan []byte, 1)
-	m.m[signature] = ch
-	m.mu.Unlock()
-
-	return Matcher{
-		C:   ch,
-		sig: signature,
-		m:   m,
-	}, nil
+	ch := make(chan []byte, bufSize)
+	m.m[signature] = incoming{expected: bufSize, c: ch}
+	return ch, nil
 }
 
-// Cleanup closes underlying channel and deletes signature from map
-func (m *Matcher) Cleanup() {
-	m.m.mu.Lock()
-	close(m.C)
-	delete(m.m.m, m.sig)
-	m.m.mu.Unlock()
+// RemoveSignature removes the signature response from map and closes the channel.
+func (m *Match) RemoveSignature(signature any) {
```

Comment:
Please add test 😄

Reply:
I would still like a test

Reply:
```go
	ch, err = match.Set("masterblaster", 1)
	require.NoError(t, err)
	match.RemoveSignature("masterblaster")
	garbage := <-ch // Should be closed and super slipery
	require.Empty(t, garbage)
```

Was [added in to satisfy the test](https://github.com/thrasher-corp/gocryptotrader/pull/1581/commits/ee8a35c42d8e1fec5070600356acd4905e23970a), so if the signature wasn't removed this would hang indefinitely. Would you like me to break it out further into its own test function?

Reply:
Yes please

Reply:
[done](https://github.com/thrasher-corp/gocryptotrader/pull/1581/commits/44a3461eb09d183914463795d5695db5f83082b5)